hack/generated/pkg/reconcilers/azure_deployment_reconciler.go

```diff
@@ -0,0 +1,869 @@
+)
+
+// TODO: I think we will want to pull some of this back into the Generic Controller so that it happens
+// TODO: for all resources
+
+const (
+	// TODO: Delete these later in favor of something in status?
+	DeploymentIdAnnotation   = "deployment-id.infra.azure.com"
+	DeploymentNameAnnotation = "deployment-name.infra.azure.com"
+	ResourceStateAnnotation  = "resource-state.infra.azure.com"
+	ResourceIdAnnotation     = "resource-id.infra.azure.com"
+	ResourceErrorAnnotation  = "resource-error.infra.azure.com"
+	ResourceSigAnnotationKey = "resource-sig.infra.azure.com"
+	// PreserveDeploymentAnnotation is the key which tells the applier to keep or delete the deployment
+	PreserveDeploymentAnnotation = "x-preserve-deployment"
+)
+
+// TODO: Do we actually want this at the controller level or this level?
+const GenericControllerFinalizer = "generated.infra.azure.com/finalizer"
+
+type CreateOrUpdateAction string
+
+const (
+	CreateOrUpdateActionNoAction          = CreateOrUpdateAction("NoAction")
+	CreateOrUpdateActionManageOwnership   = CreateOrUpdateAction("ManageOwnership")
+	CreateOrUpdateActionBeginDeployment   = CreateOrUpdateAction("BeginDeployment")
+	CreateOrUpdateActionMonitorDeployment = CreateOrUpdateAction("MonitorDeployment")
+)
+
+type DeleteAction string
+
+const (
+	DeleteActionBeginDelete   = DeleteAction("BeginDelete")
+	DeleteActionMonitorDelete = DeleteAction("MonitorDelete")
+)
+
+// TODO: I'm pretty sure we can come up with a way to drop the action parameter here
+type CreateOrUpdateActionFunc = func(ctx context.Context, action CreateOrUpdateAction) (ctrl.Result, error)
+type DeleteActionFunc = func(ctx context.Context, action DeleteAction) (ctrl.Result, error)
+
+var _ genruntime.Reconciler = &AzureDeploymentReconciler{}
+
+type AzureDeploymentReconciler struct {
+	obj                  genruntime.MetaObject
+	log                  logr.Logger
+	recorder             record.EventRecorder
+	ARMClient            armclient.Applier
+	KubeClient           *kubeclient.Client
+	ResourceResolver     *armresourceresolver.Resolver
+	CreateDeploymentName func(obj metav1.Object) (string, error)
+}
+
+func NewAzureDeploymentReconciler(
+	metaObj genruntime.MetaObject,
+	log logr.Logger,
+	armClient armclient.Applier,
+	eventRecorder record.EventRecorder,
+	kubeClient *kubeclient.Client,
+	resourceResolver *armresourceresolver.Resolver,
+	createDeploymentName func(obj metav1.Object) (string, error)) genruntime.Reconciler {
+
+	return &AzureDeploymentReconciler{
+		obj:                  metaObj,
+		log:                  log,
+		recorder:             eventRecorder,
+		ARMClient:            armClient,
+		CreateDeploymentName: createDeploymentName,
+		KubeClient:           kubeClient,
+		ResourceResolver:     resourceResolver,
+	}
+}
+
+func (r *AzureDeploymentReconciler) CreateOrUpdate(ctx context.Context) (ctrl.Result, error) {
+	action, actionFunc, err := r.DetermineCreateOrUpdateAction()
```

Comment:
Since both `action` and `actionFunc` are returned from the same factory method, we shouldn't need to pass the `action` back in again, the func can already have that value captured. 

We do need to return `action` from the method, so that we can do the right logging further on, but that's all.

Reply:
Agree - there's a tight coupling between methods and the corresponding action enum as well.

Have fixed this.